src/constraint/encoding.ts

```diff
@@ -208,18 +195,35 @@ export const ENCODING_CONSTRAINTS: EncodingConstraintModel[] = [
       return true; // other channel is irrelevant to this constraint
     }
   }, {
-    name: 'dataTypeMatchesScaleType',
+    name: 'dataTypeAndFunctionMatchesScaleType',
```

Comment:
We order the constraints mostly by alphabetical order to facilitate finding, except when a constraint is expensive (mostly spec constraints, not encoding constraints), we put them at the end.  Can you move all the new constraints to preserve alphabetical order?